usr.bin/diff/diff.1
.\" $OpenBSD: diff.1,v 1.47 2015/11/24 19:35:41 jmc Exp $
.\" $NetBSD: diff.1,v 1.2 2025/12/04 22:04:58 wiz Exp $
.\"
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)diff.1	8.1 (Berkeley) 6/30/93
.\"
.Dd July 26, 2024
.Dt DIFF 1
.Os
.Sh NAME
.Nm diff
.Nd differential file and directory comparator
.Sh SYNOPSIS
.Nm diff
.Op Fl aBbdipTtw
.Oo
.Fl c | e | f |
.Fl n | q | u | y
.Oc
.Op Fl Fl brief
.Op Fl Fl changed-group-format Ar GFMT
.Op Fl Fl ed
.Op Fl Fl expand-tabs
.Op Fl Fl forward-ed
.Op Fl Fl ignore-all-space
.Op Fl Fl ignore-case
.Op Fl Fl ignore-space-change
.Op Fl Fl initial-tab
.Op Fl Fl minimal
.Op Fl Fl no-dereference
.Op Fl Fl no-ignore-file-name-case
.Op Fl Fl normal
.Op Fl Fl rcs
.Op Fl Fl show-c-function
.Op Fl Fl starting-file
.Op Fl Fl speed-large-files
.Op Fl Fl strip-trailing-cr
.Op Fl Fl text
.Op Fl Fl unified
.Op Fl I Ar pattern | Fl Fl ignore-matching-lines Ar pattern
.Op Fl F Ar pattern | Fl Fl show-function-line Ar pattern
.Op Fl L Ar label | Fl Fl label Ar label
.Ar file1 file2
.Nm diff
.Op Fl aBbdilpTtw
.Op Fl I Ar pattern | Fl Fl ignore-matching-lines Ar pattern
.Op Fl F Ar pattern | Fl Fl show-function-line Ar pattern
.Op Fl L Ar label | Fl Fl label Ar label
.Op Fl Fl brief
.Op Fl Fl changed-group-format Ar GFMT
.Op Fl Fl ed
.Op Fl Fl expand-tabs
.Op Fl Fl forward-ed
.Op Fl Fl ignore-all-space
.Op Fl Fl ignore-case
.Op Fl Fl ignore-space-change
.Op Fl Fl initial-tab
.Op Fl Fl minimal
.Op Fl Fl no-dereference
.Op Fl Fl no-ignore-file-name-case
.Op Fl Fl normal
.Op Fl Fl paginate
.Op Fl Fl rcs
.Op Fl Fl show-c-function
.Op Fl Fl speed-large-files
.Op Fl Fl starting-file
.Op Fl Fl strip-trailing-cr
.Op Fl Fl text
.Fl C Ar number | Fl Fl context Ar number
.Ar file1 file2
.Nm diff
.Op Fl aBbdiltw
.Op Fl I Ar pattern | Fl Fl ignore-matching-lines Ar pattern
.Op Fl Fl brief
.Op Fl Fl changed-group-format Ar GFMT
.Op Fl Fl ed
.Op Fl Fl expand-tabs
.Op Fl Fl forward-ed
.Op Fl Fl ignore-all-space
.Op Fl Fl ignore-case
.Op Fl Fl ignore-space-change
.Op Fl Fl initial-tab
.Op Fl Fl minimal
.Op Fl Fl no-dereference
.Op Fl Fl no-ignore-file-name-case
.Op Fl Fl normal
.Op Fl Fl paginate
.Op Fl Fl rcs
.Op Fl Fl show-c-function
.Op Fl Fl speed-large-files
.Op Fl Fl starting-file
.Op Fl Fl strip-trailing-cr
.Op Fl Fl text
.Fl D Ar string | Fl Fl ifdef Ar string
.Ar file1 file2
.Nm diff
.Op Fl aBbdilpTtw
.Op Fl I Ar pattern | Fl Fl ignore-matching-lines Ar pattern
.Op Fl F Ar pattern | Fl Fl show-function-line Ar pattern
.Op Fl L Ar label | Fl Fl label Ar label
.Op Fl Fl brief
.Op Fl Fl changed-group-format Ar GFMT
.Op Fl Fl ed
.Op Fl Fl expand-tabs
.Op Fl Fl forward-ed
.Op Fl Fl ignore-all-space
.Op Fl Fl ignore-case
.Op Fl Fl ignore-space-change
.Op Fl Fl initial-tab
.Op Fl Fl minimal
.Op Fl Fl no-dereference
.Op Fl Fl no-ignore-file-name-case
.Op Fl Fl normal
.Op Fl Fl paginate
.Op Fl Fl rcs
.Op Fl Fl show-c-function
.Op Fl Fl speed-large-files
.Op Fl Fl starting-file
.Op Fl Fl strip-trailing-cr
.Op Fl Fl text
.Fl U Ar number | Fl Fl unified Ar number
.Ar file1 file2
.Nm diff
.Op Fl aBbdilNPprsTtw
.Oo
.Fl c | e | f |
.Fl n | q | u
.Oc
.Op Fl Fl brief
.Op Fl Fl changed-group-format Ar GFMT
.Op Fl Fl context
.Op Fl Fl ed
.Op Fl Fl expand-tabs
.Op Fl Fl forward-ed
.Op Fl Fl ignore-all-space
.Op Fl Fl ignore-case
.Op Fl Fl ignore-space-change
.Op Fl Fl initial-tab
.Op Fl Fl minimal
.Op Fl Fl new-file
.Op Fl Fl no-dereference
.Op Fl Fl no-ignore-file-name-case
.Op Fl Fl normal
.Op Fl Fl paginate
.Op Fl Fl rcs
.Op Fl Fl recursive
.Op Fl Fl report-identical-files
.Op Fl Fl show-c-function
.Op Fl Fl speed-large-files
.Op Fl Fl strip-trailing-cr
.Op Fl Fl text
.Op Fl Fl unidirectional-new-file
.Op Fl Fl unified
.Op Fl I Ar pattern | Fl Fl ignore-matching-lines Ar pattern
.Op Fl F Ar pattern | Fl Fl show-function-line Ar pattern
.Bk -words
.Op Fl L Ar label | Fl Fl label Ar label
.Op Fl S Ar name | Fl Fl starting-file Ar name
.Op Fl X Ar file | Fl Fl exclude-from Ar file
.Op Fl x Ar pattern | Fl Fl exclude Ar pattern
.Ek
.Ar dir1 dir2
.Nm diff
.Op Fl aBbditwW
.Op Fl Fl expand-tabs
.Op Fl Fl ignore-all-space
.Op Fl Fl ignore-blank-lines
.Op Fl Fl ignore-case
.Op Fl Fl minimal
.Op Fl Fl no-dereference
.Op Fl Fl no-ignore-file-name-case
.Op Fl Fl strip-trailing-cr
.Op Fl Fl suppress-common-lines
.Op Fl Fl text
.Op Fl Fl width
.Fl y | Fl Fl side-by-side
.Ar file1 file2
.Sh DESCRIPTION
The
.Nm
utility compares the contents of
.Ar file1
and
.Ar file2
and writes to the standard output the list of changes necessary to
convert one file into the other.
No output is produced if the files are identical.
.Pp
Output options (mutually exclusive):
.Bl -tag -width Ds
.It Fl C Ar number Fl Fl context Ar number
Like
.Fl c
but produces a diff with
.Ar number
lines of context.
.It Fl c
Produces a diff with 3 lines of context.
With
.Fl c
the output format is modified slightly:
the output begins with identification of the files involved and
their creation dates and then each change is separated
by a line with fifteen
.Li * Ns 's .
The lines removed from
.Ar file1
are marked with
.Sq \&-\ \& ;
those added to
.Ar file2
are marked
.Sq +\ \& .
Lines which are changed from one file to the other are marked in
both files with
.Sq !\ \& .
Changes which lie within 3 lines of each other are grouped together on
output.
.It Fl D Ar string Fl Fl ifdef Ar string
Creates a merged version of
.Ar file1
and
.Ar file2
on the standard output, with C preprocessor controls included so that
a compilation of the result without defining
.Ar string
is equivalent to compiling
.Ar file1 ,
while defining
.Ar string
will yield
.Ar file2 .
.It Fl e Fl Fl ed
Produces output in a form suitable as input for the editor utility,
.Xr ed 1 ,
which can then be used to convert file1 into file2.
.Pp
Extra commands are added to the output when comparing directories with
.Fl e ,
so that the result is a
.Xr sh 1
script for converting text files which are common to the two directories
from their state in
.Ar dir1
to their state in
.Ar dir2 .
.It Fl f Fl Fl forward-ed
Identical output to that of the
.Fl e
flag, but in reverse order.
It cannot be digested by
.Xr ed 1 .
.It Fl n
Produces a script similar to that of
.Fl e ,
but in the opposite order and with a count of changed lines on each
insert or delete command.
This is the form used by rcsdiff.
.It Fl q Fl Fl brief
Just print a line when the files differ.
Does not output a list of changes.
.It Fl U Ar number Fl Fl unified Ar number
Like
.Fl u
but produces a diff with
.Ar number
lines of context.
.It Fl u
Produces a
.Em unified
diff with 3 lines of context.
A unified diff is similar to the context diff produced by the
.Fl c
option.
However, unlike with
.Fl c ,
all lines to be changed (added and/or removed) are present in
a single section.
.It Fl y Fl Fl side-by-side
Output in two columns with a marker between them.
The marker can be one
of the following:
.Pp

doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.3205 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 11.0 to NetBSD 12.0:

	pvscsi(4): Add a port of the FreeBSD driver for the VMware PV SCSI
		controller found in VMware ESXi. [skrll 20250804]
	posix_openpt(3): Support O_CLOFORK. [gutteridge 20250806]
	sh(1): cd "" is now an error, as required by POSIX 2024. [kre 20250812]
	macppc: Add drivebay(4), a driver to control the Xserve G4's drive bays
		via sysctl. [macallan 20250812]
	macppc: Added support for fan control on Xserve G4. [macallan 20250820]
	acpi(4): Updated ACPICA to 20250807. [christos 20250824]
	binutils: Import 2.45 and switch amd64 to it  [christos 20250826]
	gffb(4): Add backlight control to the GeForce framebuffer driver.
		[macallan 20250830]
	man(1): Add -l flag to parse argument as file name.
		[christos 20250902]
	gdb: Import 16.3 and switch everyone to it [christos 20250902]
	openpam(3): update to 20250531 (zingiber) [christos 20250903]
	openldap: Import 2.6.10. [christos 20250904]
	unbound(8): Import 1.23.1. [christos 20250906]
	kernel: Add new memory barrier for paravirtualized drivers
		requiring store-before-load ordering. [riastradh 20250906]
	macppc: Add driver for the LED bar found on Xserve G4,
		controls lights depending on CPU usage. [macallan 20250908]
	m68k: Change softfloat build to libc/softfloat. [nat 20250910]
	fsck_ffs(8): Expand flags with -a and -i. [perseant 20250914]
	ctype(3): New environment variable LIBC_ALLOWCTYPEABUSE.
		[riastradh 20250915]
	dumplfs(8): Add -n flag to allow dumping of a given inode from an
		inode block, or every inode in a block. [perseant 20250914]
	lfs: Add working in-kernel roll forward. [perseant 20250917]
	kernel: Add a shared function to query the common properties used
		for configuring an Ethernet address. [thorpej 20251004]
	gffb(4): Add backlight control for NVIDIA GeForce FX 5200 Go.
		[macallan 20251006]
	hypotl(3): Accommodate infinity for softfloat. [nat 20251007]
	arm, aarch64: Switch to GCC 14. [mrg 20251007]
	sparc, sparc64: Switch to GCC 14. [mrg 20251007]
	hppa: Switch to GCC 14. [mrg 20251007]
	x86: Switch to GCC 14. [mrg 20251007]
	ucom(4): Add support for console over USB-to-serial adapter
		[manu 20251009]
	libc: Merge extended worker threads for POSIX asynchronous I/O
		(GSoC project by Ethan Miller). [christos 20251010]
	ssh(1), sshd(8): Import OpenSSH 10.2. [christos 20251011]
	gpt(8): Add various riscv-specific partition types. [thorpej 20251011]
	fsck_lfs(8): Enable checking LFS64 volumes. [perseant 20251012]
	riscv: Handle XTheadMae vs. Svpbmt, allowing boot on Allwinner
		Nezha board. [thorpej 20251012]
	viaide(4): Add UDMA support for VT8233C IDE controller.
		[andvar 20251017]
	m68k: New port: mac68ksf - Macintosh 68k softfloat. [nat 20251020]
	sc16is7xx(4): NXP SC16IX7xx family of UART chips that are behind
		a I2C or SPI bus.  [brad 20251024]
	compat_linux(8): Add support for semtimedop(2), from Ricardo Branco.
		[christos 20251026]
	ftp(1): Add "connect" as a synonym for "open". [lukem 20251028]
	wscons(4): Add support for xterm-style 256 colour escapes,
		from Crystal Kolipe. [nia 20251102]
	lfs: Add in-kernel cleaner. [perseant 20251106]
	aac(4): Support Adaptec RAID 5445, 5805 and 5085. [riz 20251107]

usr.bin/diff/diff.c
/*	$OpenBSD: diff.c,v 1.67 2019/06/28 13:35:00 deraadt Exp $	*/

/*
 * Copyright (c) 2003 Todd C. Miller <Todd.Miller@courtesan.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Sponsored in part by the Defense Advanced Research Projects
 * Agency (DARPA) and Air Force Research Laboratory, Air Force
 * Materiel Command, USAF, under agreement number F39502-99-1-0512.
 */

#include <sys/cdefs.h>
#include <sys/stat.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <limits.h>

#include "diff.h"
#include "xmalloc.h"

bool	 lflag, Nflag, Pflag, rflag, sflag, Tflag, cflag;
bool	 ignore_file_case, suppress_common;
int	 diff_format, diff_context, status;
int	 width = 130;
char	*start, *ifdefname, *diffargs, *label[2];
char	*ignore_pats, *most_recent_pat;
char	*group_format = NULL;
const char	*add_code, *del_code;
struct stat stb1, stb2;
struct excludes *excludes_list;
regex_t	 ignore_re, most_recent_re;

#define	OPTIONS	"0123456789aBbC:cdD:efF:HhI:iL:lnNPpqrS:sTtU:uwW:X:x:y"
enum {
	OPT_TSIZE = CHAR_MAX + 1,
	OPT_STRIPCR,
	OPT_IGN_FN_CASE,
	OPT_NO_IGN_FN_CASE,
	OPT_NORMAL,
	OPT_HELP,
	OPT_HORIZON_LINES,
	OPT_CHANGED_GROUP_FORMAT,
	OPT_SUPPRESS_COMMON,
	OPT_VERSION,
};

static struct option longopts[] = {
	{ "text",			no_argument,		0,	'a' },
	{ "ignore-space-change",	no_argument,		0,	'b' },
	{ "context",			optional_argument,	0,	'C' },
	{ "ifdef",			required_argument,	0,	'D' },
	{ "minimal",			no_argument,		0,	'd' },
	{ "ed",				no_argument,		0,	'e' },
	{ "forward-ed",			no_argument,		0,	'f' },
	{ "show-function-line",		required_argument,	0,	'F' },
	{ "speed-large-files",		no_argument,		NULL,	'H' },
	{ "ignore-blank-lines",		no_argument,		0,	'B' },
	{ "ignore-matching-lines",	required_argument,	0,	'I' },
	{ "ignore-case",		no_argument,		0,	'i' },
	{ "paginate",			no_argument,		NULL,	'l' },
	{ "label",			required_argument,	0,	'L' },
	{ "new-file",			no_argument,		0,	'N' },
	{ "rcs",			no_argument,		0,	'n' },
	{ "unidirectional-new-file",	no_argument,		0,	'P' },
	{ "show-c-function",		no_argument,		0,	'p' },
	{ "brief",			no_argument,		0,	'q' },
	{ "recursive",			no_argument,		0,	'r' },
	{ "report-identical-files",	no_argument,		0,	's' },
	{ "starting-file",		required_argument,	0,	'S' },
	{ "expand-tabs",		no_argument,		0,	't' },
	{ "initial-tab",		no_argument,		0,	'T' },
	{ "unified",			optional_argument,	0,	'U' },
	{ "ignore-all-space",		no_argument,		0,	'w' },
	{ "width",			required_argument,	0,	'W' },
	{ "exclude",			required_argument,	0,	'x' },
	{ "exclude-from",		required_argument,	0,	'X' },
	{ "side-by-side",		no_argument,		NULL,	'y' },
	{ "ignore-file-name-case",	no_argument,		NULL,	OPT_IGN_FN_CASE },
	{ "horizon-lines",		required_argument,	NULL,	OPT_HORIZON_LINES },
	{ "no-ignore-file-name-case",	no_argument,		NULL,	OPT_NO_IGN_FN_CASE },
	{ "normal",			no_argument,		NULL,	OPT_NORMAL },
	{ "strip-trailing-cr",		no_argument,		NULL,	OPT_STRIPCR },
	{ "changed-group-format",	required_argument,	NULL,	OPT_CHANGED_GROUP_FORMAT},
	{ "suppress-common-lines",	no_argument,		NULL,	OPT_SUPPRESS_COMMON },
	{ NULL,				0,			0,	'\0'}
};

static void checked_regcomp(char const *, regex_t *);
static void usage(void) __dead;
static void conflicting_format(void) __dead;
static void push_excludes(char *);
static void push_ignore_pats(char *);
static void read_excludes_file(char *file);
static void set_argstr(char **, char **);
static char *splice(char *, char *);

int
main(int argc, char **argv)
{
	char *ep, **oargv;
	long  l;
	int   ch, dflags, lastch, gotstdin, prevoptind, newarg;

	oargv = argv;
	gotstdin = 0;
	dflags = 0;
	lastch = '\0';
	prevoptind = 1;
	newarg = 1;
	diff_context = 3;
	diff_format = D_UNSET;
#define	FORMAT_MISMATCHED(type)	\
	(diff_format != D_UNSET && diff_format != (type))
	while ((ch = getopt_long(argc, argv, OPTIONS, longopts, NULL)) != -1) {
		switch (ch) {
		case '0': case '1': case '2': case '3': case '4':
		case '5': case '6': case '7': case '8': case '9':
			if (newarg)
				usage();	/* disallow -[0-9]+ */
			else if (lastch == 'c' || lastch == 'u')
				diff_context = 0;
			else if (!isdigit(lastch) || diff_context > INT_MAX / 10)
				usage();
			diff_context = (diff_context * 10) + (ch - '0');
			break;
		case 'a':
			dflags |= D_FORCEASCII;
			break;
		case 'b':
			dflags |= D_FOLDBLANKS;
			break;
		case 'C':
		case 'c':
			if (FORMAT_MISMATCHED(D_CONTEXT))
				conflicting_format();
			cflag = true;
			diff_format = D_CONTEXT;
			if (optarg != NULL) {
				l = strtol(optarg, &ep, 10);
				if (*ep != '\0' || l < 0 || l >= INT_MAX)
					usage();
				diff_context = (int)l;
			}
			break;
		case 'd':
			dflags |= D_MINIMAL;
			break;
		case 'D':
			if (FORMAT_MISMATCHED(D_IFDEF))
				conflicting_format();
			diff_format = D_IFDEF;
			ifdefname = optarg;
			break;
		case 'e':
			if (FORMAT_MISMATCHED(D_EDIT))
				conflicting_format();
			diff_format = D_EDIT;
			break;
		case 'f':
			if (FORMAT_MISMATCHED(D_REVERSE))
				conflicting_format();
			diff_format = D_REVERSE;
			break;
		case 'H':
			/* ignore but needed for compatibility with GNU diff */
			break;
		case 'h':
			/* silently ignore for backwards compatibility */
			break;
		case 'B':
			dflags |= D_SKIPBLANKLINES;
			break;
		case 'F':
			if (dflags & D_PROTOTYPE)
				conflicting_format();
			dflags |= D_MATCHLAST;
			most_recent_pat = xstrdup(optarg);
			break;
		case 'I':
			push_ignore_pats(optarg);
			break;
		case 'i':
			dflags |= D_IGNORECASE;
			break;
		case 'L':
			if (label[0] == NULL)
				label[0] = optarg;
			else if (label[1] == NULL)
				label[1] = optarg;
			else
				usage();
			break;
		case 'l':
			lflag = true;
			break;
		case 'N':
			Nflag = true;
			break;
		case 'n':
			if (FORMAT_MISMATCHED(D_NREVERSE))
				conflicting_format();
			diff_format = D_NREVERSE;
			break;
		case 'p':
			if (dflags & D_MATCHLAST)
				conflicting_format();
			dflags |= D_PROTOTYPE;
			break;
		case 'P':
			Pflag = true;
			break;
		case 'r':
			rflag = true;
			break;
		case 'q':
			if (FORMAT_MISMATCHED(D_BRIEF))
				conflicting_format();
			diff_format = D_BRIEF;
			break;
		case 'S':
			start = optarg;
			break;
		case 's':
			sflag = true;
			break;
		case 'T':
			Tflag = true;
			break;
		case 't':
			dflags |= D_EXPANDTABS;
			break;
		case 'U':
		case 'u':
			if (FORMAT_MISMATCHED(D_UNIFIED))
				conflicting_format();
			diff_format = D_UNIFIED;
			if (optarg != NULL) {
				l = strtol(optarg, &ep, 10);
				if (*ep != '\0' || l < 0 || l >= INT_MAX)
					usage();
				diff_context = (int)l;
			}
			break;
		case 'w':
			dflags |= D_IGNOREBLANKS;
			break;
		case 'W':
			if (optarg != NULL) {
				l = strtol(optarg, &ep, 10);
				if (*ep != '\0' || l < 0 || l >= INT_MAX)
					usage();
				width = (int)l;
			} else {
				usage();
			}
			break;
		case 'X':
			read_excludes_file(optarg);
			break;
		case 'x':
			push_excludes(optarg);
			break;
		case 'y':
			if (FORMAT_MISMATCHED(D_SIDEBYSIDE))
				conflicting_format();
			diff_format = D_SIDEBYSIDE;
			break;
		case OPT_CHANGED_GROUP_FORMAT:
			if (FORMAT_MISMATCHED(D_GFORMAT))
				conflicting_format();
			diff_format = D_GFORMAT;
			group_format = optarg;
			break;
		case OPT_HORIZON_LINES:
			break; /* XXX TODO for compatibility with GNU diff3 */
		case OPT_IGN_FN_CASE:
			ignore_file_case = true;
			break;
		case OPT_NO_IGN_FN_CASE:
			ignore_file_case = false;
			break;
		case OPT_NORMAL:
			if (FORMAT_MISMATCHED(D_NORMAL))
				conflicting_format();
			diff_format = D_NORMAL;
			break;
		case OPT_STRIPCR:
			dflags |= D_STRIPCR;
			break;
		case OPT_SUPPRESS_COMMON:
			suppress_common = 1;
			break;
		default:
			usage();
			break;
		}
		lastch = ch;
		newarg = optind != prevoptind;
		prevoptind = optind;
	}
	if (diff_format == D_UNSET && (dflags & D_PROTOTYPE) != 0)
		diff_format = D_CONTEXT;
	if (diff_format == D_UNSET)
		diff_format = D_NORMAL;
	argc -= optind;
	argv += optind;

	/*
	 * Do sanity checks, fill in stb1 and stb2 and call the appropriate
	 * driver routine.  Both drivers use the contents of stb1 and stb2.
	 */
	if (argc != 2)
		usage();
	checked_regcomp(ignore_pats, &ignore_re);
	checked_regcomp(most_recent_pat, &most_recent_re);
	if (strcmp(argv[0], "-") == 0) {
		fstat(STDIN_FILENO, &stb1);
		gotstdin = 1;
	} else if (stat(argv[0], &stb1) != 0) {
		if (!Nflag || errno != ENOENT)
			err(2, "%s", argv[0]);
		dflags |= D_EMPTY1;
		memset(&stb1, 0, sizeof(struct stat));
	}

	if (strcmp(argv[1], "-") == 0) {
		fstat(STDIN_FILENO, &stb2);
		gotstdin = 1;
	} else if (stat(argv[1], &stb2) != 0) {
		if (!Nflag || errno != ENOENT)
			err(2, "%s", argv[1]);
		dflags |= D_EMPTY2;
		memset(&stb2, 0, sizeof(stb2));
		stb2.st_mode = stb1.st_mode;
	}

	if (dflags & D_EMPTY1 && dflags & D_EMPTY2){
		warn("%s", argv[0]);
		warn("%s", argv[1]);
		exit(2);
	}

	if (stb1.st_mode == 0)
		stb1.st_mode = stb2.st_mode;

	if (gotstdin && (S_ISDIR(stb1.st_mode) || S_ISDIR(stb2.st_mode)))
		errx(2, "can't compare - to a directory");
	set_argstr(oargv, argv);
	if (S_ISDIR(stb1.st_mode) && S_ISDIR(stb2.st_mode)) {
		if (diff_format == D_IFDEF)
			errx(2, "-D option not supported with directories");
		diffdir(argv[0], argv[1], dflags);
	} else {
		if (S_ISDIR(stb1.st_mode)) {
			argv[0] = splice(argv[0], argv[1]);
			if (stat(argv[0], &stb1) == -1)
				err(2, "%s", argv[0]);
		}
		if (S_ISDIR(stb2.st_mode)) {
			argv[1] = splice(argv[1], argv[0]);
			if (stat(argv[1], &stb2) == -1)
				err(2, "%s", argv[1]);
		}
		print_status(diffreg(argv[0], argv[1], dflags), argv[0],
		    argv[1], "");
	}
	exit(status);
}

static void
checked_regcomp(char const *pattern, regex_t *comp)
{
	char buf[BUFSIZ];
	int error;

	if (pattern == NULL)
		return;

	error = regcomp(comp, pattern, REG_NEWLINE | REG_EXTENDED);
	if (error != 0) {
		regerror(error, comp, buf, sizeof(buf));
		if (*pattern != '\0')
			errx(2, "%s: %s", pattern, buf);
		else
			errx(2, "%s", buf);
	}
}

static void
set_argstr(char **av, char **ave)
{
	size_t argsize;
	char **ap;

	argsize = 4 + *ave - *av + 1;
	diffargs = xmalloc(argsize);
	strlcpy(diffargs, "diff", argsize);
	for (ap = av + 1; ap < ave; ap++) {
		if (strcmp(*ap, "--") != 0) {
			strlcat(diffargs, " ", argsize);
			strlcat(diffargs, *ap, argsize);
		}
	}
}

/*
 * Read in an excludes file and push each line.
 */
static void
read_excludes_file(char *file)
{
	FILE *fp;
	char *buf, *pattern;
	size_t len;

	if (strcmp(file, "-") == 0)
		fp = stdin;
	else if ((fp = fopen(file, "r")) == NULL)
		err(2, "%s", file);
	while ((buf = fgetln(fp, &len)) != NULL) {
		if (buf[len - 1] == '\n')
			len--;
		if ((pattern = strndup(buf, len)) == NULL)
			err(2, "xstrndup");
		push_excludes(pattern);
	}
	if (strcmp(file, "-") != 0)
		fclose(fp);
}

/*
 * Push a pattern onto the excludes list.
 */
static void
push_excludes(char *pattern)
{
	struct excludes *entry;

	entry = xmalloc(sizeof(*entry));
	entry->pattern = pattern;
	entry->next = excludes_list;
	excludes_list = entry;
}

static void
push_ignore_pats(char *pattern)
{
	size_t len;

	if (ignore_pats == NULL)
		ignore_pats = xstrdup(pattern);
	else {
		/* old + "|" + new + NUL */
		len = strlen(ignore_pats) + strlen(pattern) + 2;
		ignore_pats = xreallocarray(ignore_pats, 1, len);
		strlcat(ignore_pats, "|", len);
		strlcat(ignore_pats, pattern, len);
	}
}

void
print_status(int val, char *path1, char *path2, const char *entry)
{
	if (label[0] != NULL)
		path1 = label[0];
	if (label[1] != NULL)
		path2 = label[1];

	switch (val) {
	case D_BINARY:
		printf("Binary files %s%s and %s%s differ\n",
		    path1, entry, path2, entry);
		break;
	case D_DIFFER:
		if (diff_format == D_BRIEF)
			printf("Files %s%s and %s%s differ\n",
			    path1, entry, path2, entry);
		break;
	case D_SAME:
		if (sflag)
			printf("Files %s%s and %s%s are identical\n",
			    path1, entry, path2, entry);
		break;
	case D_MISMATCH1:
		printf("File %s%s is a directory while file %s%s is a regular file\n",
		    path1, entry, path2, entry);
		break;
	case D_MISMATCH2:
		printf("File %s%s is a regular file while file %s%s is a directory\n",
		    path1, entry, path2, entry);
		break;
	case D_SKIPPED1:
		printf("File %s%s is not a regular file or directory and was skipped\n",
		    path1, entry);
		break;
	case D_SKIPPED2:
		printf("File %s%s is not a regular file or directory and was skipped\n",
		    path2, entry);
		break;
	case D_ERROR:
		break;
	}
}

static void
usage(void)
{
	(void)fprintf(stderr,
	    "usage: diff [-aBbdilpTtw] [-c | -e | -f | -n | -q | -u] [--ignore-case]\n"
	    "            [--no-ignore-case] [--normal] [--strip-trailing-cr]\n"
	    "            [-I pattern] [-F pattern] [-L label] file1 file2\n"
	    "       diff [-aBbdilpTtw] [-I pattern] [-L label] [--ignore-case]\n"
	    "            [--no-ignore-case] [--normal] [--strip-trailing-cr]\n"
	    "            [-F pattern] -C number file1 file2\n"
	    "       diff [-aBbdiltw] [-I pattern] [--ignore-case] [--no-ignore-case]\n"
	    "            [--normal] [--strip-trailing-cr] -D string file1 file2\n"
	    "       diff [-aBbdilpTtw] [-I pattern] [-L label] [--ignore-case]\n"
	    "            [--no-ignore-case] [--normal] [--strip-trailing-cr]\n"
	    "            [-F pattern] -U number file1 file2\n"
	    "       diff [-aBbdilNPprsTtw] [-c | -e | -f | -n | -q | -u] [--ignore-case]\n"
	    "            [--no-ignore-case] [--normal] [-I pattern] [-L label]\n"
	    "            [-F pattern] [-S name] [-X file] [-x pattern] dir1 dir2\n"
	    "       diff [-aBbditwW] [--expand-tabs] [--ignore-all-space]\n"
	    "            [--ignore-blank-lines] [--ignore-case] [--minimal]\n"
	    "            [--no-ignore-file-name-case] [--strip-trailing-cr]\n"
	    "            [--suppress-common-lines] [--text] [--width]\n"
	    "            -y | --side-by-side file1 file2\n");

	exit(2);
}

static void
conflicting_format(void)
{

	fprintf(stderr, "error: conflicting output format options.\n");
	usage();
}

static char *
splice(char *dir, char *path)
{
	char *tail, *buf;
	size_t dirlen;

	dirlen = strlen(dir);
	while (dirlen != 0 && dir[dirlen - 1] == '/')
	    dirlen--;
	if ((tail = strrchr(path, '/')) == NULL)
		tail = path;
	else
		tail++;
	xasprintf(&buf, "%.*s/%s", (int)dirlen, dir, tail);
	return (buf);
}
